ouroboros-network/src/Ouroboros/Network/Tracers/OnlyForTracer.hs

```diff
@@ -0,0 +1,99 @@
+{-# LANGUAGE Rank2Types #-}
+
+-- | See 'OnlyForTracer'
+--
+-- Intended for qualified import.
+--
+-- > import           Ouroboros.Network.Tracers.OnlyForTracer (OnlyForTracer)
+-- > import qualified Ouroboros.Network.Tracers.OnlyForTracer as OnlyForTracer
+module Ouroboros.Network.Tracers.OnlyForTracer
+  ( OnlyForTracer (..)
+  , PolyRun (..)
+  , contramap
+  , mkPolyRun
+  , mkRun
+    -- * Convenience combinators
+  , contramapM
+  , lift
+  , mkTracerWith
+  , mkTracerWithPolyRun
+  , pure
+  , traceWith
+  ) where
+
+import           Prelude hiding (Applicative (..))
+import qualified Prelude
+
+import           Control.Tracer (Tracer (..))
+import qualified Control.Tracer as Tracer
+
+-- | An opaque box holding data that can only be accessed by a 'Tracer'
+--
+-- Note: In sufficient monads, the intended use can be subverted in order to
+-- leak the contained data outside of tracer invocations, but that should be
+-- caught and avoided during code review.
+newtype OnlyForTracer a = UnsafeOnlyForTracer {
+    -- | This is only safe to use _inside_ of another 'OnlyForTracer'
+    -- computation
+    unsafeRunOnlyForTracer :: a
+  }
+
+instance Functor OnlyForTracer where
+  fmap f (UnsafeOnlyForTracer a) = UnsafeOnlyForTracer (f a)
+
+instance Prelude.Applicative OnlyForTracer where
+  pure = pure
```

Comment:
I see `pure` is defined at the top level; why not just 'UnsafeOnlyForTracer' here?

Reply:
I switched to the following, which avoids the awkward `pure = pure` line.

```
instance Prelude.Applicative OnlyForTracer where
  pure = UnsafeOnlyForTracer
  UnsafeOnlyForTracer f <*> UnsafeOnlyForTracer a = UnsafeOnlyForTracer (f a)

-- | The only safe use of 'UnsafeOnlyForTracer'
pure :: a -> OnlyForTracer a
pure = Prelude.pure
```